util/argo/argo.go

```diff
@@ -455,3 +460,56 @@ func getDestinationServer(ctx context.Context, db db.ArgoDB, clusterName string)
 	}
 	return servers[0], nil
 }
+
+func GetAppVirtualProject(proj *argoappv1.AppProject, projLister applicationsv1.AppProjectLister, settingsManager *settings.SettingsManager) (*argoappv1.AppProject, error) {
+	gps, err := settingsManager.GetGlobalProjectsSettings()
+	if err != nil {
+		log.Warnf("Failed to get global project settings")
+		return proj, err
+	}
+	virtualProj := proj.DeepCopy()
+
+	for _, gp := range gps {
+		selector, err := metav1.LabelSelectorAsSelector(&gp.LabelSelector)
+		if err != nil {
+			break
+		}
+		//Get projects which match the label selector, then see if proj is a match
+		projList0, err := projLister.AppProjects(proj.Namespace).List(selector)
+		if err != nil {
+			break
+		}
+		var matchMe bool
+		for _, item := range projList0 {
+			if item.Name == proj.Name {
+				matchMe = true
+				break
+			}
+		}
+		if !matchMe {
+			break
+		}
+		//If proj is a match for this global project setting, then merge with the global project
+		globalProj, err := projLister.AppProjects(proj.Namespace).Get(gp.ProjectName)
+		if err != nil {
+			break
+		}
+		virtualProj = mergeVirtualProject(virtualProj, globalProj)
+	}
+	return virtualProj, err
```

Comment:
I think you want to `return virtualProj, nil`.   Errors can be overwritten in the loop.  It does not look like you are preservation or reporting the error(s) anyway.

Reply:
@wtam2018 Thank you very much for the detailed review!